File: src/data/courses/dsa/trie/counting-dictionary/word-break-trie/code/python/word_break_trie.py
# ============================================================
# Solution 1: Brute Force Recursion — O(2^N)
# ============================================================
class Solution1:
    def wordBreak(self, s, wordDict):
        words = set(wordDict)
        def canBreak(start):
            if start == len(s): return True
            for end in range(start+1, len(s)+1):
                if s[start:end] in words and canBreak(end): return True
            return False
        return canBreak(0)

# ============================================================
# Solution 2: Trie + DP — O(N*L)
# ============================================================
class TrieNode:
    def __init__(self): self.children, self.is_end = {}, False

class Solution2:
    def wordBreak(self, s, wordDict):
        root = TrieNode()
        for word in wordDict:
            node = root
            for c in word:
                if c not in node.children: node.children[c] = TrieNode()
                node = node.children[c]
            node.is_end = True
        n = len(s)
        dp = [False] * (n + 1); dp[0] = True
        for i in range(n):
            if not dp[i]: continue
            node = root
            for j in range(i, n):
                if s[j] not in node.children: break
                node = node.children[s[j]]
                if node.is_end: dp[j+1] = True
        return dp[n]

File: src/data/courses/dsa/trie/counting-dictionary/word-break-trie/code/cpp/WordBreakTrie.cpp
// ============================================================
// Solution 1: Brute Force Recursion — try all splits
// Time: O(2^N) | Space: O(N)
// ============================================================
#include <vector>
#include <string>
#include <unordered_set>
using namespace std;

class Solution1 {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> dict(wordDict.begin(), wordDict.end());
        return canBreak(s, 0, dict);
    }
    bool canBreak(string& s, int start, unordered_set<string>& dict) {
        if (start == (int)s.size()) return true;
        for (int end = start + 1; end <= (int)s.size(); end++)
            if (dict.count(s.substr(start, end - start)) && canBreak(s, end, dict))
                return true;
        return false;
    }
};

// ============================================================
// Solution 2: Trie + DP — O(N × L)
// Time: O(N × L) where L = max word length | Space: O(N + D×L)
// ============================================================
class Solution2 {
    struct TrieNode { TrieNode* ch[26] = {}; bool isEnd = false; };
    TrieNode* root = new TrieNode();
    
    void insert(string& word) {
        TrieNode* n = root;
        for (char c : word) {
            if (!n->ch[c-'a']) n->ch[c-'a'] = new TrieNode();
            n = n->ch[c-'a'];
        }
        n->isEnd = true;
    }
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        for (auto& w : wordDict) insert(w);
        int n = s.size();
        vector<bool> dp(n + 1, false);
        dp[0] = true;
        for (int i = 0; i < n; i++) {
            if (!dp[i]) continue;
            TrieNode* node = root;
            for (int j = i; j < n; j++) {
                if (!node->ch[s[j]-'a']) break; // No more words match
                node = node->ch[s[j]-'a'];
                if (node->isEnd) dp[j + 1] = true; // Word found!
            }
        }
        return dp[n];
    }
};
